src/base_alloc/base_alloc.c

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (C) 2024 Intel Corporation
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+*/
+
+#include <assert.h>
+
+#include "base_alloc.h"
+#include "base_alloc_os_mem.h"
+#include "utils_common.h"
+#include "utils_concurrency.h"
+
+// pool size of the base allocator,
+// it will be reallocated if we run out of memory.
+#define BA_POOL_SIZE (10 * ba_os_get_page_size())
+
+// alignment of the base allocator
+#define MEMORY_ALIGNMENT (8)
+
+typedef struct ba_mem_chunk_t ba_mem_chunk_t;
+
+// memory chunk of size 'aligned_size'
+struct ba_mem_chunk_t {
+    ba_mem_chunk_t *next;
+    char user_data[];
+};
+
+struct umf_ba_pool_meta_t {
+    size_t aligned_size;       // size of all memory chunks in this pool
+    os_mutex_t *free_lock;     // lock of free_list
+    ba_mem_chunk_t *free_list; // list of free chunks
+};
+
+struct umf_ba_pool_t {
+    struct umf_ba_pool_meta_t metadata;
+    char data[];
+};
+
+// ba_divide_memory_into_chunks - divide given memory into chunks of aligned_size and add them to the free_list
+static void ba_divide_memory_into_chunks(umf_ba_pool_t *pool, void *ptr,
+                                         size_t size) {
+    assert(pool->metadata.free_list == NULL);
+    assert(size > pool->metadata.aligned_size);
+
+    char *pool_memory = ptr;
+    size_t size_left = size;
+
+    ba_mem_chunk_t *current_chunk = (ba_mem_chunk_t *)pool_memory;
+    ba_mem_chunk_t *prev_chunk = current_chunk;
+
+    while (size_left >= pool->metadata.aligned_size) {
+        current_chunk = (ba_mem_chunk_t *)pool_memory;
+        prev_chunk->next = current_chunk;
+
+        pool_memory += pool->metadata.aligned_size;
+        size_left -= pool->metadata.aligned_size;
+        prev_chunk = current_chunk;
+    }
+
+    current_chunk->next = NULL;
+    pool->metadata.free_list = ptr; // address of the first chunk
+}
+
+umf_ba_pool_t *umf_ba_create(size_t size) {
+    umf_ba_pool_t *pool = (umf_ba_pool_t *)ba_os_alloc(BA_POOL_SIZE);
+    if (!pool) {
+        return NULL;
+    }
+
+    char *pool_memory = (char *)&pool->data;
+    size_t size_left = BA_POOL_SIZE - sizeof(struct umf_ba_pool_meta_t);
+
+    // align size of a chunk to MEMORY_ALIGNMENT bytes
+    pool->metadata.aligned_size =
+        align_size(sizeof(ba_mem_chunk_t) + size, MEMORY_ALIGNMENT);
+
+    // allocate and init free_lock
+    pool->metadata.free_lock = util_mutex_init(pool_memory);
+    if (!pool->metadata.free_lock) {
+        return NULL;
+    }
+
+    pool_memory += util_mutex_get_size(); // free_lock is here
+    size_left -= util_mutex_get_size();   // for free_lock
+
+    util_mutex_lock(pool->metadata.free_lock);
+    pool->metadata.free_list = NULL;
+    ba_divide_memory_into_chunks(pool, pool_memory, size_left);
```

Comment:
Do we need to actually divide all the memory eagerly at the beginning?

I think we could get rid of ba_divide_memory_into_chunks and just do 'bump' allocation for the first `pool_size / aligned_size` allocations. That is, have an `allocated_size` variable which would initially be 0. Every time we allocate, if `allocated_size + aligned_size` is <= `pool_size` then we just return `pool->base_address + allocated_size`. If `allocated_size + aligned_size` >= `pool_size` we look at the free list.

And the free list is only populated on `umf_ba_free`.

Not sure if this is worth doing, tough - the overhead of `ba_os_alloc` probably makes the ba_divide_memory_into_chunks unnoticable.

Reply:
I mean, I'd rather have the code be simpler and preemptively split it in a cold path, then have to special-case the hot path.

Reply:
It will happen on each allocation that needs to call `ba_os_alloc` (when free list is empty). But yeah, it might actually be faster anyway since we have less atomics.

Reply:
@igchor is it resolved now?